### ISSUE TITLE
control Type creation with Decorator registration

### ISSUE THREAD

**Issue by redoced** [simpleinjector/SimpleInjector#61]
When registering a Decorator I would like to use a factory that not only returns the type but the actual instance to use for decoration. So basically I would like to create the Decorator myself.

Why?
There's that neat nuget package EventSourceProxy that allows you to create a ETW proxy from an instance. Since this is done by calling a method, I see no way to do this with registering a decorator in SimpleInjector.

So basically I would like to do this:

    container.RegisterDecorator(typeof(IInterface), (IInterface decorateThisInstance) => TracingProxy.CreateWithActivityScope<IInterface>(decorateThisInstance));

Is this possible?

**Comment by qujck** [comment]
You can probably do it with a standard registration:

```csharp
public interface IInterface { }
public class Implementation : IInterface { }

container.Register<Implementation>();
container.Register<IInterface>(() => 
    TracingProxy.CreateWithActivityScope<IInterface>(
        container.GetInstance<Implementation>()));
```

**Comment by redoced** [comment]
That's how I do it for now, but it doesn't seem right. What if I wanted to decorate another decorator?

**Comment by dotnetjunkie** [comment]
What do you mean by "decorate another decorator"? Can you give an example?

**Comment by redoced** [comment]
http://simpleinjector.readthedocs.org/en/latest/advanced.html#decorators

I basically just mean to chain up decorators. One in that chain should be my TracingProxy.

**Comment by dotnetjunkie** [comment]
I would say in that case you just do this:

``` c#
container.Register<Implementation>();
container.Register<IInterface>(() =>
    new LoggingInterfaceDecorator(
        container.GetInstance<ILogger>(),
        TracingProxy.CreateWithActivityScope<IInterface>(
            container.GetInstance<Implementation>())));
```
As you already noticed, there is no `RegisterDecorator<TService>(Func<TService, TService> decoratorApplier)` method. There are two reasons for the absence of such method:

 1. The `RegisterDecorator` overloads are in the first place meant for decorating generic types. It is extremely difficult to apply decorators to generic types, without built-in support, while it is quite easy to do so for non-generic types. When registering open generic types, it is impossible to supply such delegate, because you would.
 2. Simple Injector favors being able to verify object graphs for you. Anytime to supply a delegate to Simple Injector however, you make the verification and diagnostic system blind.

That's why there is no `RegisterDecorator<TService>(Func<TService, TService> decoratorApplier)` and it will probably never be added.

**Comment by abatishchev** [comment]
+1 to have `RegisterDecorator<TService>(Func<TService, TService> decoratorApplier)`
